### ISSUE TITLE
Complex FormPanel data?

### ISSUE THREAD

**Issue by wardellbagby** [rjaros/kvision#70]
Let's say I have a data class that looks like:

```
@Serializable
data class MyData(val myInt: Int, val myNestedData: NestedData)
@Serializable
data class MyNestedData(val myString: String)
```

Is it possible to have `FormPanel` represent the `MyData` class directly? Or do I need to unwind this data class such that it looks like:

```
@Serializable
data class MyFormPanelData(val myInt: Int, val myNestedData_myString: String)
```

**Comment by rjaros** [comment]
It's not possible. KVision keeps internal form data model in a Map, and uses kotlinx.serialization [Mapper](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/runtime_usage.md#mapper) to serialize/deserialize data.

**Comment by wardellbagby** [comment]
So I see. Are there any plans to support functionality like this one day?

**Comment by rjaros** [comment]
If the nested class has only one property (like in your example) it probably could be done with some effort. But if there are more properties I see no way to reference them and add form controls for them (`FormPanel.add` takes a field reference as a parameter).

The only thing I can imagine is to implement nested forms - such nested classes could be automatically mapped to nested forms. But it would be hard and would take a lot of work. It's not impossible but I'm afraid it won't be done anytime soon.

**Comment by wardellbagby** [comment]
I've been scratching my head around this for a while, and I'd really like to avoid having to create a brand new unwrapped data class solely for driving a `FormPanel`. Especially because I feel like it's just asking for issues keeping these two different classes in sync.

Obviously, nested forms are out of the question, and so this follow-up question is a bit out of scope of this issue but I thought I'd ask anyway. Feel free to ignore me, but if you had a data class with complex values, something like:

```
data class MyDataClass(val name: String, val description: NestedDataClass1, val items: List<NestedDataClass2>)
```

how would you go about structuring a UI to handle this data structure, assuming you want it to be editable by a user? Can you think of any examples that might already exist that I could look at? 

I appreciate you taking the time to help me out with this!

**Comment by rjaros** [comment]
I would probably try to use tabs. One tab for every item in the list. If the list is short - horizontal tabs, if it can be long - vertical tabs. Switching tabs would show a form panel bound to the one of nested data classes in the list. It should be relatively easy to write functions to populate data from the model to all the form panels in the tabs and read the data back after some user event.
If the nested forms are relatively small, you can just show them all at once without tabs - the code would be very similar.

**Comment by robert-cronin** [comment]
I've also been wondering if we need serializable constraint for forms as well, is this another case where we can get rid of kotlinx dependency? if you can use it in JavaScript without serialization, why couldn't you do the same thing in KotlinJS?

**Comment by rjaros** [comment]
`Form<K>` class uses `Map<String, FormControl>` internally to keep the data and it needs to create `K` class somehow (in the `getData()` method). We cannot just create an instance of generic `K` class in Kotlin nor in JS. 
Early implementations of KVision's `Form` class required model class properties to be delegated to Map and required additional factory function in the constructor (see how it looked: https://github.com/rjaros/kvision-examples/blob/4c8553f91f9bb2f7b5df0255c56e1aec14963bc9/showcase/src/main/kotlin/com/example/FormTab.kt). Then I found out kotlinx.serialization had a `Mapper` class, which allowed to create `K` instance automagically, as long as it's serializable. So I've integrated this solution, and the serialization requirement is a lot better than the previous one. I'm not sure it can be dropped without sacrificing the functionality, but I'm open to any suggestions.
